### PR TITLE
[EPMCDLAB-1788]: added automation scheduler for stopping/terminating clusters in case of inactivity

### PR DESCRIPTION
**Pull request by ochaparin**

### CODE REVIEW

File: services/provisioning-service/src/main/java/com/epam/dlab/backendapi/core/response/handlers/CheckInactivityClustersCallbackHandler.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+		final String fileContent = new String(content);
+		log.debug("Got file {} while waiting for UUID {}, check inactivity clusters response: {}", fileName, uuid,
+				fileContent);
+
+		final JsonNode treeNode = MAPPER.readTree(fileContent);
+		final String status = treeNode.get(STATUS_FIELD).textValue();
+		CheckInactivityClustersStatusDTO checkInactivityClustersStatusDTO;
+		if ("ok".equals(status)) {
+			final JsonNode clustersNode = treeNode.get(CLUSTERS_FIELD);
+			ObjectReader reader = MAPPER.readerFor(new TypeReference<List<EnvResource>>() {
```
Comment (bhliva) [inline services/provisioning-service/src/main/java/com/epam/dlab/backendapi/core/response/handlers/CheckInactivityClustersCallbackHandler.java]: please extract method for ok/failed status

File: services/provisioning-service/src/main/java/com/epam/dlab/backendapi/core/response/handlers/CheckInactivityClustersCallbackHandler.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.ObjectReader;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.ws.rs.core.Response;
+import java.util.Collections;
+import java.util.List;
+
+@Slf4j
+public class CheckInactivityClustersCallbackHandler implements FileHandlerCallback {
```
Comment (bhliva) [inline services/provisioning-service/src/main/java/com/epam/dlab/backendapi/core/response/handlers/CheckInactivityClustersCallbackHandler.java]: Rename CheckInactivityClustersCallbackHandler -> CheckInactivityClusterCallbackHandler

File: services/self-service/src/main/java/com/epam/dlab/backendapi/resources/aws/ComputationalResourceAws.java [hunk 3]
```diff
@@ -177,6 +178,29 @@ public Response start(@Auth UserInfo userInfo,
...
+	 *
+	 * @param userInfo                user info.
+	 * @param exploratoryName         name of exploratory.
+	 * @param computationalName       name of computational resource.
+	 * @param checkInactivityRequired true/false.
+	 * @return 200 OK if operation is successfully triggered
+	 */
+	@POST
+	@Path("/{exploratoryName}/{computationalName}/inactivity")
+	public Response updateCheckInactivity(@Auth UserInfo userInfo,
```
Comment (bhliva) [inline services/self-service/src/main/java/com/epam/dlab/backendapi/resources/aws/ComputationalResourceAws.java]: rename to updateInactivity

File: services/self-service/src/main/java/com/epam/dlab/backendapi/resources/azure/ComputationalResourceAzure.java [hunk 4]
```diff
@@ -150,4 +150,27 @@ public Response start(@Auth UserInfo userInfo,
...
+	 *
+	 * @param userInfo                user info.
+	 * @param exploratoryName         name of exploratory.
+	 * @param computationalName       name of computational resource.
+	 * @param checkInactivityRequired true/false.
+	 * @return 200 OK if operation is successfully triggered
+	 */
+	@POST
+	@Path("/{exploratoryName}/{computationalName}/inactivity")
+	public Response updateCheckInactivity(@Auth UserInfo userInfo,
```
Comment (bhliva) [inline services/self-service/src/main/java/com/epam/dlab/backendapi/resources/azure/ComputationalResourceAzure.java]: rename to updateInactivity

File: services/self-service/src/main/java/com/epam/dlab/backendapi/resources/callback/CheckInactiveClustersCallback.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+import javax.ws.rs.Path;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.UriInfo;
+
+@Path("/infrastructure/check_inactivity/clusters/callback")
+@Consumes(MediaType.APPLICATION_JSON)
+@Slf4j
+public class CheckInactiveClustersCallback {
```
Comment (bhliva) [inline services/self-service/src/main/java/com/epam/dlab/backendapi/resources/callback/CheckInactiveClustersCallback.java]: rename CheckInactiveClusterCallback

File: services/self-service/src/main/java/com/epam/dlab/backendapi/resources/gcp/ComputationalResourceGcp.java [hunk 6]
```diff
@@ -176,6 +177,29 @@ public Response start(@Auth UserInfo userInfo,
...
+	 *
+	 * @param userInfo                user info.
+	 * @param exploratoryName         name of exploratory.
+	 * @param computationalName       name of computational resource.
+	 * @param checkInactivityRequired true/false.
+	 * @return 200 OK if operation is successfully triggered
+	 */
+	@POST
+	@Path("/{exploratoryName}/{computationalName}/inactivity")
+	public Response updateCheckInactivity(@Auth UserInfo userInfo,
```
Comment (bhliva) [inline services/self-service/src/main/java/com/epam/dlab/backendapi/resources/gcp/ComputationalResourceGcp.java]: -> updateInactivity and method should be PUT

File: services/self-service/src/main/java/com/epam/dlab/backendapi/service/ComputationalService.java [hunk 7]
```diff
@@ -62,4 +63,11 @@ void updateComputationalsReuploadKeyFlag(String user, List<UserInstanceStatus> e
 
 	Optional<UserComputationalResource> getComputationalResource(String user, String exploratoryName,
 																 String computationalName);
+
+	void stopOrTerminateClustersByCondition(CheckInactivityClustersStatusDTO dto);
```
Comment (bhliva) [inline services/self-service/src/main/java/com/epam/dlab/backendapi/service/ComputationalService.java]: rename to stopClustersByCondition

File: services/dlab-model/src/main/java/com/epam/dlab/dto/computational/CheckInactivityClustersStatus.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+ *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  * See the License for the specific language governing permissions and
+ *  * limitations under the License.
+ *
+ */
+package com.epam.dlab.dto.computational;
+
+import java.util.Arrays;
+
+public enum CheckInactivityClustersStatus {
```
Comment (bhliva) [inline services/dlab-model/src/main/java/com/epam/dlab/dto/computational/CheckInactivityClustersStatus.java]: rename to CheckInactivityClusterStatus

File: services/dlab-model/src/main/java/com/epam/dlab/dto/computational/CheckInactivityClustersStatusDTO.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+import com.epam.dlab.dto.StatusBaseDTO;
+import com.epam.dlab.dto.status.EnvResource;
+import com.google.common.base.MoreObjects;
+import lombok.Getter;
+
+import java.util.List;
+
+@Getter
+public class CheckInactivityClustersStatusDTO extends StatusBaseDTO<CheckInactivityClustersStatusDTO> {
```
Comment (bhliva) [inline services/dlab-model/src/main/java/com/epam/dlab/dto/computational/CheckInactivityClustersStatusDTO.java]: rename to CheckInactivityClusterStatusDTO

File: services/provisioning-service/src/main/java/com/epam/dlab/backendapi/core/response/handlers/CheckInactivityClustersCallbackHandler.java (hunk 1 above)
Comment (ochaparin) [reply services/provisioning-service/src/main/java/com/epam/dlab/backendapi/core/response/handlers/CheckInactivityClustersCallbackHandler.java]: done

File: services/provisioning-service/src/main/java/com/epam/dlab/backendapi/core/response/handlers/CheckInactivityClustersCallbackHandler.java (hunk 2 above)
Comment (ochaparin) [reply services/provisioning-service/src/main/java/com/epam/dlab/backendapi/core/response/handlers/CheckInactivityClustersCallbackHandler.java]: done

File: services/self-service/src/main/java/com/epam/dlab/backendapi/resources/aws/ComputationalResourceAws.java (hunk 3 above)
Comment (ochaparin) [reply services/self-service/src/main/java/com/epam/dlab/backendapi/resources/aws/ComputationalResourceAws.java]: done

File: services/self-service/src/main/java/com/epam/dlab/backendapi/resources/azure/ComputationalResourceAzure.java (hunk 4 above)
Comment (ochaparin) [reply services/self-service/src/main/java/com/epam/dlab/backendapi/resources/azure/ComputationalResourceAzure.java]: done

File: services/self-service/src/main/java/com/epam/dlab/backendapi/resources/callback/CheckInactiveClustersCallback.java (hunk 5 above)
Comment (ochaparin) [reply services/self-service/src/main/java/com/epam/dlab/backendapi/resources/callback/CheckInactiveClustersCallback.java]: done

File: services/self-service/src/main/java/com/epam/dlab/backendapi/resources/gcp/ComputationalResourceGcp.java (hunk 6 above)
Comment (ochaparin) [reply services/self-service/src/main/java/com/epam/dlab/backendapi/resources/gcp/ComputationalResourceGcp.java]: done

File: services/self-service/src/main/java/com/epam/dlab/backendapi/service/ComputationalService.java (hunk 7 above)
Comment (ochaparin) [reply services/self-service/src/main/java/com/epam/dlab/backendapi/service/ComputationalService.java]: done

File: services/dlab-model/src/main/java/com/epam/dlab/dto/computational/CheckInactivityClustersStatus.java (hunk 8 above)
Comment (ochaparin) [reply services/dlab-model/src/main/java/com/epam/dlab/dto/computational/CheckInactivityClustersStatus.java]: done

File: services/dlab-model/src/main/java/com/epam/dlab/dto/computational/CheckInactivityClustersStatusDTO.java (hunk 9 above)
Comment (ochaparin) [reply services/dlab-model/src/main/java/com/epam/dlab/dto/computational/CheckInactivityClustersStatusDTO.java]: done

**Pull request merged**